### ISSUE TITLE
Improve sorting of includes

### ISSUE THREAD

**Issue by mwoehlke-kitware** [uncrustify/uncrustify#1929]
I'm... not really in love with the current implementation of sorting includes. Having only two categories, in particular, is limiting, and it's not really optimal for the implementation of `sorting.cpp`, either, especially in light of #1852.

What I'd like to see is something more like clang-format, where instead of options, there is a *directive* to add include regexes, each of which has an open-ended priority value. These should be processed in declaration order, so that more specific regexes can be matched first, even if they have a lower priority (this is important for some use cases).

These changes may depend on #1815.

I'd also like to see:

- A regex to mutate the name of the file being processed in order to determine its "name". Possibly this would include specifying which match group to extract.
- Ability to include the result of the above in other regexes.
- (Maybe) Ability to specify a "gap", such that include groups with priority less than the gap don't get blank lines between them.

**Comment by mwoehlke-kitware** [comment]
One "gotcha" I just realized I don't know if we support... includes in other PP stuff need to be kept separate. Example:
```c++
#include <foo.h>
#if HAVE_BAR_H
#include <bar.h>
#endif
```

Ideally I would like to be able to sort these, but I think we would want an option whether to interleave them, or put them at the end of their respective blocks (and optionally with a blank line first). Interleaving would need to be able to deal with cases like:
```c++
#include <z.h>
#if HAVE_BAR
#include <d.h>
#endif
#if HAVE_FOO
#include <c.h>
#include <f.h>
#endif
#if HAVE_BAR
#include <e.h>
#endif
```
...which should become:
```c++
#if HAVE_FOO
#include <c.h>
#endif
#if HAVE_BAR
#include <d.h>
#include <e.h>
#endif
#if HAVE_FOO
#include <f.h>
#endif
#include <z.h>
```

**Comment by arneforlie** [comment]
Another issue with the current implementation is that the includes are sorted case insensitively. So a list of #includes such as:
```
#include <AClass.h>
#include <TheClass.h>
#include <SomeClass.h>
#include <iostream>
#include <vector>
```
turns into:
```
#include <AClass.h>
#include <iostream>
#include <SomeClass.h>
#include <TheClass.h>
#include <vector>
```

As an intermediate solution, perhaps the option could be changed to:
mod_sort_include = ignore | case | nocase

**Comment by mihaipopescu** [comment]
@mwoehlke-kitware About PP directives, the gotcha is for uncrustify itself, not only for sorting includes.

As a general comment, PR #959 introduced the _clang inspired_ regex categories. The idea is great but this was the only way I found to implement them by having 3 categories. Of course we can just increase the number of categories if you need more but I'm not sure about the rest of your feature request.

@arneforlie about case-sensitive sorting, it should be case sensitive. As per #959, the sorting prioritization is done as follows:
> The sorting order is controlled by the priority of the category first followed by a string comparison with a higher priority on lowercase versus uppercase.

If this is not the case, it's a bug.

**Comment by arneforlie** [comment]
Perhaps I'm misunderstanding this, but I do not consider the result of my test case to be a case sensitive sort. If lowercase should precede uppercase, I would have expected:
```
#include <iostream>
#include <vector>
#include <AClass.h>
#include <SomeClass.h>
#include <TheClass.h>
```
But, no, that's not what I'm getting.

**Comment by mihaipopescu** [comment]
@arneforlie Can you paste your config as well?

**Comment by mihaipopescu** [comment]
Actaully, please report this as another issue. We don't want to get too much off-topic on this issue.
